### ISSUE TITLE
Proposal for a go-redis like high level api

### ISSUE THREAD

**Issue by 418Coffee** [rueian/rueidis#18]
Hi, I plan to use ``rueidis`` for some of my next projects, and I must say that I like your ''bottom-up" approach. But, currently writing commands is quite cumbersome. I would like to propose a design for a high-level API similar to what [go-redis](https://github.com/go-redis/redis#quickstart) has.

Rueidis' "Getting Started" section uses this:
```go
// SET key val NX
c.Do(ctx, c.B().Set().Key("key").Value("val").Nx().Build()).Error()
// GET key
c.Do(ctx, c.B().Get().Key("key").Build()).ToString()
```

What if we could turn that into:
```go
// SET key val NX
c.SetNX(ctx, "key", "val").Error()
// GET key
c.Get(ctx, "key").ToString()
```

Well maybe it's not hard to do so, this was the most straightforward thing I could think of. We basically wrap cmds.Builder:
```go
func (c *singleClient) SetNX(ctx context.Context, key, value string) (resp RedisResult) {
	cmd := c.B().Set().Key(key).Value(value).Nx().Build()
	resp = c.Do(ctx, cmd)
	cmds.Put(cmd.CommandSlice())
	return resp
}

func (c *singleClient) Get(ctx context.Context, key string) (resp RedisResult) {
	cmd := c.B().Get().Key(key).Build()
	resp = c.Do(ctx, cmd)
	cmds.Put(cmd.CommandSlice())
	return resp
}
```

The issue comes when trying to implement things like caching, how could the user specify to use client-side caching?

We could create a separate function for each command?
Each command would have two functions, ``*`` and ``*Cache`` (where ``*`` is the command).
```go
func (c *singleClient) SetNXCache(ctx context.Context, key, value string) (resp RedisResult) 

func (c *singleClient) GetCache(ctx context.Context, key, value string) (resp RedisResult) 
```
Or perhaps by some kind of hacky optional boolean?
```go
func (c *singleClient) SetNXCache(ctx context.Context, key, value string, cache ...bool) (resp RedisResult) 

func (c *singleClient) GetCache(ctx context.Context, key, value string, cache ...bool) (resp RedisResult) 
```
Or still something different?

Let me know what you think about this, If we can decide on a concise design I wouldn't mind implementing it.

**Comment by rueian** [comment]
Hi @418Coffee,

Thank you for proposing the design of a high-level API so that I can have a discussion on this topic.

I like the idea. Actually, I have been considering implementing exactly the same go-redis's [`Cmdable`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) interface for a while, including its response structs such as [`StringCmd`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L860) and [`IntCmd`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L598).

In my imagination, this will be a separated adapter in another package, for example:

```golang
package main

import (
	"context"
	"fmt"
	"time"
	"github.com/rueian/rueidis"
	"github.com/rueian/rueidis/rueidiscompat"
)

func main() {
	client, err := rueidis.NewClient(rueidis.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	defer client.Close()

	compat := rueidiscompat.NewAdapter(client)
	ok, err := compat.SetNX(ctx, "key", "val", time.Second).Result()
}

```

In terms of client-side caching, since it requires the user to provide an explicit TTL, I think it would be better structured like this:

```golang
ok, err := compat.Cache(time.Second).Get(ctx, "key").Result()
```

**Comment by 418Coffee** [comment]
> I think I could implement a subset of it first. Would you mind sharing what commands you would like to use in your next projects?

My original statement about using ``rueidis`` in future projects was general, I don't know what exact commands I'll use.

If I understand correctly, [go-redis](https://github.com/go-redis/redis) only implements [``Cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) once on the [``cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L416) type. And then uses it on all client types, Ring, Cluster, etc. In my previous proposal I only implemented high-level commands for the [``singleClient``](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/client.go#L10) struct. I agree that a generic interface and type like [``Cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) and [``cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L416) is a better approach. I assume that ``NewAdapter`` would wrap the [``singleClient``](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/client.go#L10), or whatever connection type, and implement the [``Cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) interface. I also like the proposal of chaining the command function by prepending the ``Cache`` method. 

However, I don't really see the need for the [``StringCmd``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L860), [``IntCmd``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L598), etc. structs. Doesn't [``RedisResult``](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/message.go#L74) take care of that? Albeit in a more generic approach.

Finally, wouldn't the actual implementation of [``Cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) not be much different from what I proposed? That is, just wrapping the command building?

Here is some pseudo-code:
```go
func (c *Compat) Get(key string) (resp RedisResult) {
        cmd := c.client.B().Get().Key(key).Build()
	resp = c.client.Do(ctx, cmd)
	cmd.Put(cmd.CommandSlice())
	return resp
}
```

**Comment by rueian** [comment]
> I assume that `NewAdapter` would wrap the [`singleClient`](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/client.go#L10), or whatever connection type, and implement the [`Cmdable`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) interface.

I think the `NewAdapter` would be better wrapping the [`rueidis.Client`](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/rueidis.go#L96) interface, and then all client structs (`singleClient`, `clusterClient` and `sentinelClient`) can be applied by the adapter.

> However, I don't really see the need for the [`StringCmd`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L860), [`IntCmd`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L598), etc. structs. Doesn't [`RedisResult`](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/message.go#L74) take care of that? Albeit in a more generic approach.

Since the proposal aims to implement a go-redis like high-level API, I think it would be better to make the API to be almost identical to go-redis if possible. This could help others migrate their codebase to rueidis.

> Finally, wouldn't the actual implementation of [Cmdable](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) be similar to what I proposed? That is, just wrapping the command building?

I believe the actual implementation would be similar to what you proposed.

**Comment by 418Coffee** [comment]
> I think the NewAdapter would be better wrapping the [rueidis.Client](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/rueidis.go#L96) interface, and then all client structs (singleClient, clusterClient and sentinelClient) can be applied by the adapter.

My bad, this is exactly what I meant.

I played around a bit with mimicking [go-redis'](https://github.com/go-redis/redis) [``Cmdable``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) interface and the [``StringCmd``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L860), [``IntCmd``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L598), etc. structs. It doesn't seem too difficult, perhaps just tedious, like you mentioned. The biggest difference is that [go-redis](https://github.com/go-redis/redis) a the ``readReply`` function on each of its ``*Cmd`` structs that parses the redis response with respect to what type the struct wraps. So for example when a [``BoolCmd``](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/command.go#L802) is used, the response is checked for "1" or "OK". If we want similar behaviour, I suggest we wrap [``RedisResult``](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/message.go#L74), it gives us a lot of handy conversion methods. But where do we check/parse the response and set the value of the type the struct wraps? In the constructor method of the ``*Cmd`` struct?

**Comment by rueian** [comment]
> I suggest we wrap [`RedisResult`](https://github.com/rueian/rueidis/blob/fc1b39c10530afa8172d3d079b9780724fd407ca/message.go#L74), it gives us a lot of handy conversion methods.

Sure, it is a good idea.

> But where do we check/parse the response and set the value of the type the struct wraps?

I guess it will be similar to this:

```golang
type BoolCmd struct {
	result rueidis.RedisResult
}

func (cmd *BoolCmd) Val() (v bool) {
	v, _ = cmd.Result()
	return
}

func (cmd *BoolCmd) Result() (v bool, err error) {
	v, err = cmd.result.AsBool()
	if rueidis.IsRedisNil(err) {
		return false, nil
	}
	return
}
```

**Comment by rueian** [comment]
> Are you sure you want me to open a PR? I haven't implemented all functions of [`Cmdable`](https://github.com/go-redis/redis/blob/730afbcffb93760e8a36cc06cfe55ab102b693a7/commands.go#L81) yet...

It is ok. I don't expect we have a full `Cmdable` implementation on day one either. 
We can leave those unfinished functions in a TODO comment and let the package be compilable.

> Alright, so perhaps we create something like a `CacheCmdable` interface that defines all of the cacheable functions? And then implement that with a `CacheCompat` struct?

Yes, that is exactly what I imagined.

**Comment by 418Coffee** [comment]
[Done!](https://github.com/rueian/rueidis/pull/19)

Shall we leave this issue if we need to discuss implementation details in the future?

**Comment by rueian** [comment]
Sure!

**Comment by rueian** [comment]
Hi @418Coffee,

I have updated some changes at https://github.com/rueian/rueidis/pull/20

**Comment by rueian** [comment]
Actually, commands will be recycled by the underlying rueidis.Client implementations. Therefore we don’t need to do that here.

**Comment by 418Coffee** [comment]
Of course 😅, thanks for correcting me.

I'll try to implement more functions this week.

**Comment by 418Coffee** [comment]
Hey! I Implemented some more functions already...
Do you want me to make a pull request for each batch of functions I've implemented? Or wait till I'm done with implementing all functions and make one big pull request?

**Comment by rueian** [comment]
@418Coffee,

Thanks a lot!

I personally prefer that we can do this in an incremental manner, but PR in any size is welcome.

**Comment by 418Coffee** [comment]
#21 merges into [rueidiscompat-2](https://github.com/rueian/rueidis/tree/rueidiscompat-2), you can check and make changes/additions and then merge into [master](https://github.com/rueian/rueidis/tree/master).

**Comment by rueian** [comment]
It looks great! I merged it and I only changed some naming issues.

**Comment by 418Coffee** [comment]
> I only changed some naming issues.
Yeah, I saw that today too, I guess my snake_case personality took over for a second 😅. 

With #22 I also changed some names to more idiomatic ones.
Same as #21, it merges into [rueidiscompat-2](https://github.com/rueian/rueidis/tree/rueidiscompat-2), you can check and make changes/additions and then merge into [master](https://github.com/rueian/rueidis/tree/master).

**Comment by 418Coffee** [comment]
Same old story, #24 merges into [rueidiscompat-2](https://github.com/rueian/rueidis/tree/rueidiscompat-2), you can check and make changes/additions and then merge into [master](https://github.com/rueian/rueidis/tree/master).

**Comment by 418Coffee** [comment]
With #26 all methods of ``Cmdable`` should now be implemented 🎉... 

Do you perhaps have a list of all cacheable commands? So I can start with defining the ``CacheCmdable`` interface.

**Comment by 418Coffee** [comment]
Hey, this is the idea i've come up with for handling cacheable commands:

```go
type CacheCmdable interface {
	Get(ctx context.Context, key string) *StringCmd
}

type CacheCompat struct {
	c *cacheCompat
}

type cacheCompat struct {
	client rueidis.Client
	ttl    time.Duration
}

func NewCacheAdapter(client rueidis.Client) *CacheCompat {
	return &CacheCompat{c: &cacheCompat{client: client}}
}

func (c *CacheCompat) Cache(ttl time.Duration) CacheCmdable {
	c.c.ttl = ttl
	return c.c
}

func (c *cacheCompat) Get(ctx context.Context, key string) *StringCmd {
	cmd := c.client.B().Get().Key(key).Cache()
	resp := c.client.DoCache(ctx, cmd, c.ttl)
	return newStringCmd(resp)
}

func main() {
        ctx := context.Background()
	client, _ := rueidis.NewClient(rueidis.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	defer client.Close()
	compat := NewCacheAdapter(client)
	res, _:= compat.Cache(time.Second).Get(ctx, "key").Result()
}
```

``CacheCmdable`` defines all cacheable commands, ``cacheCompat`` implements those commands using ``DoCache``. Then, ``CacheCompat`` wraps ``cacheCompat`` and defines a single function ``Cache``, which sets the TTL on ``cacheCompat`` and returns ``cacheCompat``. This way users must call ``Cache`` before calling any command. This approach, however, isn't thread-safe. Let me know what you think!

**Comment by rueian** [comment]
Hi @418Coffee,

Thank you for providing the design. I think we can do some improvements:

1. Instead of having a new `NewCacheAdapter(client rueidis.Client)`, I think it might be better to embed `Cache(ttl time.Duration)` in the existing `Cmdable` interface. Because I guess that most users would primarily use non-cacheable commands and only use cacheable commands under some limited scenarios.
2. We can also make `Cache(ttl time.Duration)` thread-safe by returning a new `cacheCompat` struct.

```golang
type CacheCmdable interface {
	Get(ctx context.Context, key string) *StringCmd
}

type Cmdable interface {
	Cache(ttl time.Duration) CacheCmdable
	...
}

func (c *Compat) Cache(ttl time.Duration) CacheCmdable {
	return &CacheCompat{client: c.client, ttl: ttl}
}

type CacheCompat struct {
	client rueidis.Client
	ttl    time.Duration
}

func (c *CacheCompat) Get(ctx context.Context, key string) *StringCmd {
	cmd := c.client.B().Get().Key(key).Cache()
	resp := c.client.DoCache(ctx, cmd, c.ttl)
	return newStringCmd(resp)
}

func main() {
        ctx := context.Background()
	client, _ := rueidis.NewClient(rueidis.ClientOption{InitAddress: []string{"127.0.0.1:6379"}})
	defer client.Close()
	compat := NewAdapter(client)
	compat.Set(ctx, "key", "value", 0).Result()
	res, _:= compat.Cache(time.Second).Get(ctx, "key").Result()
}
```

**Comment by 418Coffee** [comment]
>Instead of having a new NewCacheAdapter(client rueidis.Client), I think it might be better to embed Cache(ttl time.Duration) in the existing Cmdable interface.

Yes, that seems like a better way.

> We can also make Cache(ttl time.Duration) thread-safe by returning a new cacheCompat struct.

I thought of doing this, but I was afraid that at a large scale this might lead to a lot of allocations. [But I guess not...](https://gist.github.com/418Coffee/91cd5987dfefeb68ffae93bf0c1a29da)

On another note, commands built with ``c.client.B().Arbitrary`` do not have a ``Cache`` method. Can I just do ``cmds.Cacheable(cmd.Build())`` to make them cacheable?

**Comment by rueian** [comment]
> Can I just do `cmds.Cacheable(cmd.Build())` to make them cacheable?

Yes you can.

**Comment by rueian** [comment]
> I thought of doing this, but I was afraid that at a large scale this might lead to a lot of allocations.

Your concern is correct. It might lead to a lot of allocations.

However, with the inline optimization, the `&CacheCompat{}` doesn't need to escape to heap.

You can test the following with `go test -bench=BenchmarkAlloc -benchmem -gcflags="-m" .`

```golang

type CacheCompat struct {
	client rueidis.Client
	ttl    time.Duration
}

type Compat struct {
	client rueidis.Client
}

func (c *Compat) Cache(ttl time.Duration) CacheCmdable {
	return &cacheCompat{client: c.client, ttl: ttl}
}

//go:noinline
func (c *Compat) CacheNoInline(ttl time.Duration) CacheCmdable {
	return &cacheCompat{client: c.client, ttl: ttl}
}

func (c *CacheCompat) Get(ctx context.Context, key string) {}

func BenchmarkAlloc(b *testing.B) {
	c := &compat{}
	b.Run("normal", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			c.Cache(time.Second).Get(nil, "key")
		}
	})
	b.Run("noinline", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			c.CacheNoInline(time.Second).Get(nil, "key")
		}
	})
}
```

You should see the result like this:
```
pkg: github.com/rueian/rueidis/rueidiscompat
BenchmarkAlloc/normal-10         	1000000000	         0.3132 ns/op	       0 B/op	       0 allocs/op
BenchmarkAlloc/noinline-10       	66595753	        17.89 ns/op	      24 B/op	       1 allocs/op
PASS
ok  	github.com/rueian/rueidis/rueidiscompat	1.922s
```

And the output of `-gcflags="-m"` like this:
```
./cacheable_test.go:35:11: &CacheCompat{...} does not escape
```

**Comment by 418Coffee** [comment]
I did some more testing and it turns out that because we use an interface the compiler is unable to apply escape analysis to determine whether ``CacheCompat`` is too large for the stack. Thus, It must assume it's too large and allocates it on the heap.  We could turn ``NewAdapter(client rueidis.Client) Cmdable`` into `` NewAdapter(client rueidis.Client) *Compat`` to fix this, but then we lose the ``Cmdable`` interface, which is a considerable part in a go-redis like API.

**Comment by rueian** [comment]
Oh, thank you for pointing it out. I didn't aware of that. But losing the `Cmdable` interface is quite unacceptable.

I purpose just losing the `CacheCmdable` interface, that is, returning `CacheCompat` struct directly in `Cache(ttl time.Duration)`:

```golang
type Cmdable interface {
	Cache(ttl time.Duration) CacheCompat
	...
}

type CacheCompat struct {
	client rueidis.Client
	ttl    time.Duration
}

type Compat struct {
	client rueidis.Client
}

func (c *Compat) Cache(ttl time.Duration) CacheCompat {  // return struct directly
	return CacheCompat{client: c.client, ttl: ttl}
}

// use struct receiver
func (c CacheCompat) Get(ctx context.Context, key string) { ... }

func BenchmarkAlloc(b *testing.B) {
	c := (Cmdable)(&Compat{})
	for i := 0; i < b.N; i++ {
		c.Cache(time.Second).Get(nil, "key")
	}
}
```
This will also result in zero allocation.

**Comment by 418Coffee** [comment]
It's settled then. Do you want me to keep the ``CacheCmdable`` interface for reference, or shall I just remove it?

**Comment by rueian** [comment]
I think it is better to remove it to avoid confusion.

**Comment by 418Coffee** [comment]
Alright, #28 Implements all cacheable commands.

**Comment by rueian** [comment]
Thanks! I am pretty amazed that you can finish thousands of lines of code within just a few days.
I will find a way to set up test cases for the adapter. After that, your great work will be released.

**Comment by rueian** [comment]
Hi @418Coffee,

[rueidiscompat](https://github.com/rueian/rueidis/commit/a525ad43300a4ea316a8ca1355a46752ee8be8ea) is released in v0.0.44!

Thank you for your great work. I just did some changes to make the interface to be more close to go-redis', and fixed some bugs captured by the go-redis' test cases.

**Comment by 418Coffee** [comment]
Everything looks good, amazing!

Thank you for explaining stuff and always answering my questions quickly.
I suggest that this issue be closed, albeit for the time being (i.e. go-redis doesn't plan to make any great changes to ``Cmdable``).
